### ISSUE TITLE
Endiguer l’expansion des variables d'environnement

### ISSUE THREAD

**Issue by akarzim** [betagouv/demarches-simplifiees.fr#6934]
> À un moment il faut qu’on arrêter l’expansion des variables d'environnement... J'ai l'impression en plus que beaucoup de ces textes ont vocation à être traduits. Donc l'usage des variables d'environnent n'est pas possible. J'ai l'impression que la solution c'est :
> - passer ces valeurs dans i18n. Sur une instance, surcharger les locales.
> - passer le gros des variables d'environnement de configuration dans un fichier de configuration d'instance.
>
> — _Originally posted by @tchak in https://github.com/betagouv/demarches-simplifiees.fr/issues/6878#issuecomment-1028173970_

**Comment by kemenaran** [comment]
Je suis carrément chaud pour séparer les différentes configs dans des fichiers YAML séparés (et là où c'est pertinent mettre la conf dans Rails.application.x.config.*).

**Comment by kemenaran** [comment]
On pourrait même commencer par un `branding.yml`, qui contiendrait les éléments de branding non-localisables (nom de l'app, du logo, etc)

**Comment by tchak** [comment]
Pour moi c'est oui pour un `branding.yml` chargé via `config_for`.
Je serai aussi pour dans le futur ne plus avoir de globales et ne plus avoir d'accès au ENV dans le code de l'app. Mais j'ai l'impression que je suis assez seul là-dessus.

**Comment by akarzim** [comment]
> Je serai aussi pour dans le futur ne plus avoir de globales et ne plus avoir d'accès au ENV dans le code de l'app. Mais j'ai l'impression que je suis assez seul là-dessus.

@tchak je te soutiens à 100% sur cette approche !

**Comment by akarzim** [comment]
Ça m'a l'air prometteur !

Si je comprends bien, pour que la surcharge puisse se faire de manière transparente, les YAML ressembleraient plutôt à ceci : 

```yaml
# branding.yml
ds:
  app_name: "<%= ENV.fetch('DS_APP_NAME', 'demarches-simplifiees.fr') %>"
  contact_email: "<%= ENV.fetch('DS_CONTACT_EMAIL', 'contact@example.org') %>"
  # etc…
```

Concernant la configuration, il est tout à fait possible de mettre des conditions et de surcharger `Rails.config.x.*` au besoin, je ne vois pas de point de blocage de ce côté là. Tu as un exemple en tête ?

**Comment by kemenaran** [comment]
Non, l'idée ça serait d'overrider les variables automatiquement.

Un truc du genre :

```ruby
# config/initializers/read_env.rb

# For each env var that begins by `DS_`, override the matching setting in Rails.application.config.x.*
ENV.filter { |name, _| name.has_prefix?('DS_') }.each do |name, value|
  Rails.application.config.set('x.' + name.lowercase) = value
end
```

(En vrai il faudrait plutôt faire dans le sens inverse : énumérer les clefs de la config, et ensuite voir s'il y a une variable d'env correspondante. Mais c'est l'idée.)

**Comment by akarzim** [comment]
Ah oui, c'est une approche un peu plus « méta » mais ça marche aussi, je vois l'idée.

Ce qui serait parfait, en peaufinant cette approche, ce serait de pouvoir gérer proprement : 
- les valeurs booléennes (ex: `enabled`/`disabled`)
- les listes d'options finies (ex: `["amazon", "local", "openstack"]`)

Par exemple, ici la variable d'env vaut `enabled` mais la clé de configuration est bien un booléen : 

```ruby
config.x.france_connect.enabled = ENV.fetch("DS_FRANCE_CONNECT_ENABLED", "enabled") == "enabled"
```

Il faudrait ainsi pouvoir être en mesure de lever une exception si une variable d'environnement prend une valeur inattendue. Une piste pourrait être de se tourner vers l'une des gems [dry-rb][1], comme [dry-types][2] par exemple, pour décrire les clés et valeurs acceptées, et faire de la coercition quand nécessaire.

[1]:https://dry-rb.org/
[2]: https://dry-rb.org/gems/dry-types/1.2/
